### PR TITLE
[SPARK-17279][SQL] better error message for exceptions during ScalaUDF execution

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

If `ScalaUDF` throws exceptions during executing user code, sometimes it's hard for users to figure out what's wrong, especially when they use Spark shell. An example

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 12 in stage 325.0 failed 4 times, most recent failure: Lost task 12.3 in stage 325.0 (TID 35622, 10.0.207.202): java.lang.NullPointerException
    at line8414e872fb8b42aba390efc153d1611a12.$read$$iwC$$iwC$$iwC$$iwC$$anonfun$2.apply(<console>:40)
    at line8414e872fb8b42aba390efc153d1611a12.$read$$iwC$$iwC$$iwC$$iwC$$anonfun$2.apply(<console>:40)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
...
```

We should catch these exceptions and rethrow them with better error message, to say that the exception is happened in scala udf.

This PR also does some clean up for `ScalaUDF` and add a unit test suite for it.
## How was this patch tested?

the new test suite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 1]
```diff
@@ -1057,5 +1057,17 @@ case class ScalaUDF(
...
-  override def eval(input: InternalRow): Any = converter(f(input))
+  val npeErrorMessage = "Given UDF throws NPE during execution, please check the UDF " +
+    "to make sure it handles null parameters correctly."
+
+  override def eval(input: InternalRow): Any = {
+    val result = try {
+      f(input)
+    } catch {
+      case e: NullPointerException =>
+        throw new RuntimeException(npeErrorMessage, e)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Can we still use `NullPointerException`? NullPointerException can have a specified message. Then, you can use `setStackTrace` to set the original stacktrace.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 2]
```diff
@@ -1057,5 +1059,19 @@ case class ScalaUDF(
...
 
-  override def eval(input: InternalRow): Any = converter(f(input))
+  val npeErrorMessage = "Given UDF throws NPE during execution, please check the UDF " +
+    "to make sure it handles null parameters correctly."
+
+  override def eval(input: InternalRow): Any = {
+    val result = try {
+      f(input)
+    } catch {
+      case e: NullPointerException =>
```
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: It is a bit hacky to set stack trace like this.
npe.setStackTrace(e.getStackTrace)

If user search the code line reported in the stack trace, user may not able to find the code that matches the error message.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 3]
```diff
@@ -1039,9 +1032,18 @@ case class ScalaUDF(
...
+      s"""
+         ${ctx.boxedType(dataType)} $resultTerm = null;
+         try {
+           $resultTerm = (${ctx.boxedType(dataType)})$catalystConverterTerm.apply($getFuncResult);
+         } catch (NullPointerException e) {
+           NullPointerException npe = new NullPointerException($scalaUDF.npeErrorMessage());
+           npe.setStackTrace(e.getStackTrace());
+           throw npe;
+         }
+       """.stripMargin
```
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Where is margin?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 2 above)
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: 1. For this code branch `eval(input: InternalRow)`, the existing NPE message should be clear enough if there is a full stacktrace, and  the stack contains method of the UDF.
2. The error message you provided can be totally wrong.
"Given UDF throws NPE during execution, please check the UDF to make sure it handles null parameters correctly". 

What if NPE is not caused by parameter? prompting this message is misleading.

**Comment by clockfly** [comment]
@cloud-fan 

There are two branches to execute an UDF.
1. Call `override def eval(input: InternalRow): Any`  directly. For this code branch, I don't think there is any barrier for an user to know it is an UDF problem, as the ScalaUDF is in the stack trace.
2. Another branch is in codegen. This code branch is confusing, in my opinion. Since every different code piece from different expression is fused into one code block. When an exception is thrown, we don't know who owns that code block.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 4]
```diff
@@ -1039,9 +1033,18 @@ case class ScalaUDF(
       (convert, argTerm)
     }.unzip
 
-    val callFunc = s"${ctx.boxedType(dataType)} $resultTerm = " +
-      s"(${ctx.boxedType(dataType)})${catalystConverterTerm}" +
-        s".apply($funcTerm.apply(${funcArguments.mkString(", ")}));"
+    val getFuncResult = s"$funcTerm.apply(${funcArguments.mkString(", ")})"
+    val rethrowException = "throw new org.apache.spark.SparkException" +
+      """("Exception happens when execute user code in Scala UDF.", e);"""
```
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Exception happens when `executing` user defined function (className:  input argument type => output argument type) .
Or 
`Failed to execute user defined function (className: input argument type => output argument type)

**Comment by clockfly** [comment]
+1

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by cloud-fan** [comment]
also backport it to 2.0

**Pull request closed without merging**